### ISSUE TITLE
NNUE (ƎUИИ Efficiently Updatable Neural Network) for Go

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#56]
NNUE (ƎUИИ Efficiently Updatable Neural Network) is a sparse shallow neural network architecture that can be calculated incrementally. This has proven incredibly successful in [Chess](https://github.com/glinscott/nnue-pytorch/blob/master/docs/nnue.md) and Shoigi. Historically algorithms that works in Chess does not necessarily translate to Go (e.g. AlphaBeta pruning), but more modern neural network architectures has generalized well (e.g. MuZero).

This issue is to formally document the thoughts I've had on the possibility of designing a NNUE architecture for Go that performs competitively when compared to larger neural networks.

**Comment by Chicoryn** [comment]
The chess architecture is based on a denormalization of the features based on the kings position `(our_king_square, piece_square, piece_type, piece_color)`, and shogi uses a similar representation. The first challenge is to find a feature representation for Go that fulfills the criterias:

* Is hyper sparse, at most 2% non-zeros.
* Allows efficient incremental updates, at most `O(1)` amortized updates.

## `(our_piece1, our_piece2, piece_color)`

This is the "direct" translation of the chess (KP) representation for Go, which asks for storing all piece pairs. The problem is that this representation is neither sparse nor does it allow for fast incremental updates. Adding a piece to the board would require us to update `O(n)` cells as this affects every other piece on the board.

I am also skeptical about this representation as it is probably not expressive enough for a shallow network to evaluate determine long-range dependencies such as 

## MuZero Embedding representation

The NNUE architecture reassembles the MuZero architecture with a few additional constraints. Given this we could train the representation instead of trying to figure it out ourselves:

* The dynamics network is very shallow.
* The prediction network is very shallow.

## More...

...

**Comment by Chicoryn** [comment]
All search algorithms needs to weight the benefits or quality vs quantity of nodes explored. NNUE is a step towards quantity, where as all recent Go algorithms trend towards quality.

The reason for this divide can be observed in the number of candidate moves per game state, chess has about 35 whereas Go has about 150. This means the cost of a misstep during search is **a lot** higher in Go than in Chess, which might be the explanation as for why quality is more important in Go than in Chess.

This would suggest that NNUE _might_ not work as well in Go as it does for Chess.

**Comment by Chicoryn** [comment]
All current NNUE architectures are pure _value_ networks which outputs `0` or `1` depending on the current state in a losing state or not. This is not necessarily a problem, generalizing NNUE to output both a _policy_ and _value_ should not be impossible.

**Comment by Chicoryn** [comment]
Continuing on the topic of the _MuZero Embedding representation_.  It occurred to us that you can view the dynamics function as an RNN (or as it's more successful implementations GRU or LSTM). The advantage of this would be that RNN's are very well researched, and in general use only feedforward layers, which are fast. So gives us the following procedure:

1. Use the **representation** function to squash the **input features** to an **embedding vector**.
2. **Evaluate and predict the next move** from the **embedding vector**.
3. Set the **embedding vector** as the initial hidden state of the RNN.
4. Set the **input features** for the next few moves as the inputs of the RNN.
5. Use the **outputs** from the RNN for each timestep to **evaluate and predict the next move**.

The idea would be to have a potentially large initial **representation** function to compute the initial hidden state, and then allow the RNN to operate on the much smaller embedding (which would allow is to be very fast). A few open questions in regard to the modelling:

- Both GRU and LSTM makes excessive use of sigmoid instead of relu, do we need to change the final activation of the representation function to match this?
- What size should the embedding be?
- Does it make sense to feed the **embedding vector** to **both** the **prediction** function and as the initial hidden state of the RNN?
  - Should we feed start the RNN from the initial step, instead of skipping it?
  - Should we feed it an _empty_ input features instead?

# Experimental results

`WIP`

**Comment by Chicoryn** [comment]
# Experimental results

Experimental results of using a large representation network with a small dynamics network so far show negative results, the model is bottlenecked by the dynamics network and it's accuracy scale with the minimum of the dynamics / representation size. So there is no point in having two networks of different size.

If a dynamics network that is feed the exact game state cannot scale, then I do not think there is any reason to expect a dynamics network that is feed only the motion features would keep up so this is probably a dead end in terms of research.

See https://github.com/Chicoryn/dream-go/pull/58#issuecomment-1040451748 for the full data from the experiment.
